ARMmbed/meta-mbl
# Fix incremental build issue in image-signing.bbclass

**Initial commit message**
Fix incremental build issue in image-signing.bbclass

For: IOTMBL-1025: [meta-mbl#316] atf-warp7 recipe seems to break Warp7
Incremental Builds

The do_sign_root_hash task uses an artefact (the "root hash") left by
the do_generate_verity_metadata task (image-verity.bbclass), but the
name of that artefact has a timestamp in it (from $IMAGE_NAME). When
do_sign_root_hash runs, $IMAGE_NAME may have a different timestamp to
when do_generate_verity_metadata ran, and in that case do_sign_root_hash
fails to find the artefact.

Fortunately, after do_generate_verity_metadata creates the root hash, it
also creates a symlink to it with a name that doesn't contain the
timestamp, so use that symlink in do_sign_root_hash to avoid the
problem.

**Jenkins build**
http://jenkins.mbed-linux.arm.com/job/jh-test/10/ (`Failed with ERROR: Nothing PROVIDES 'mbl-image-production`').
http://jenkins.mbed-linux.arm.com/view/jh/job/jh-test/12/ (after rebase on master; succeeded).

**Testing done**
* Build and boot on Raspberry Pi 3.
* Build and boot on WaRP7.
* Run `do_generate_verity_metadata` and `do_sign_root_hash` manually to ensure that the timestamps when each is run are different.
  Before this change:
  ```
  jonhai01@e113079-lin:/data/work/mbed-linux/master7$ MACHINE=imx7s-warp-mbl . setup-environment
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ bitbake -c cleansstate mbl-console-image
  [output removed for brevity]
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ bitbake -c generate_verity_metadata mbl-console-image
  [output removed for brevity]
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ bitbake -c sign_root_hash mbl-console-image
  [output removed for brevity]
  ERROR: mbl-console-image-1.0-r0 do_sign_root_hash: Function failed: do_sign_root_hash (log file is located at /data/work/mbed-linux/master7/build-mbl/tmp-mbl-glibc/work/imx7s_warp_mbl-oe-linux-gnueabi/mbl-console-image/1.0-r0/temp/log.do_sign_root_hash.31489)
  ERROR: Logfile of failure stored in: /data/work/mbed-linux/master7/build-mbl/tmp-mbl-glibc/work/imx7s_warp_mbl-oe-linux-gnueabi/mbl-console-image/1.0-r0/temp/log.do_sign_root_hash.31489
  Log data follows:
  | DEBUG: Executing shell function do_sign_root_hash
  | /data/work/mbed-linux/master7/build-mbl/tmp-mbl-glibc/work/imx7s_warp_mbl-oe-linux-gnueabi/mbl-console-image/1.0-r0/deploy-mbl-console-image-image-complete/mbl-console-image-imx7s-warp-mbl-20181224094751.rootfs.root_hash.txt: No such file or directory
  | WARNING: exit code 1 from a shell command.
  | ERROR: Function failed: do_sign_root_hash (log file is located at /data/work/mbed-linux/master7/build-mbl/tmp-mbl-glibc/work/imx7s_warp_mbl-oe-linux-gnueabi/mbl-console-image/1.0-r0/temp/log.do_sign_root_hash.31489)
  ERROR: Task (/data/work/mbed-linux/master7/build-mbl/conf/../../layers/meta-mbl/recipes-core/images/mbl-console-image.bb:do_sign_root_hash) failed with exit code '1'
  NOTE: Tasks Summary: Attempted 3217 tasks of which 3216 didn't need to be rerun and 1 failed.
  NOTE: Writing buildhistory
  
  Summary: 1 task failed:
    /data/work/mbed-linux/master7/build-mbl/conf/../../layers/meta-mbl/recipes-core/images/mbl-console-image.bb:do_sign_root_hash
  Summary: There was 1 ERROR message shown, returning a non-zero exit code.
  ```
  After this change:
  ```
  jonhai01@e113079-lin:/data/work/mbed-linux/master7$ MACHINE=imx7s-warp-mbl . setup-environment
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ bitbake -c cleansstate mbl-console-image
  [output removed for brevity]
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ bitbake -c generate_verity_metadata mbl-console-image
  [output removed for brevity]
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ bitbake -c sign_root_hash mbl-console-image
  [output removed for brevity]
  NOTE: Tasks Summary: Attempted 3217 tasks of which 3216 didn't need to be rerun and all succeeded.
  NOTE: Writing buildhistory
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$ ls tmp-mbl-glibc/work/imx7s_warp_mbl-oe-linux-gnueabi/mbl-console-image/1.0-r0/deploy-mbl-console-image-image-complete/
  mbl-console-image-imx7s-warp-mbl-20181224095924.rootfs.ext4                           mbl-console-image-imx7s-warp-mbl.ext4
  mbl-console-image-imx7s-warp-mbl-20181224095924.rootfs.hash_tree.bin                  mbl-console-image-imx7s-warp-mbl.hash_tree.bin
  mbl-console-image-imx7s-warp-mbl-20181224095924.rootfs.manifest                       mbl-console-image-imx7s-warp-mbl.manifest
  mbl-console-image-imx7s-warp-mbl-20181224095924.rootfs.root_hash.txt                  mbl-console-image-imx7s-warp-mbl.root_hash.txt
  mbl-console-image-imx7s-warp-mbl-20181224095924.rootfs.verity_header_information.txt  mbl-console-image-imx7s-warp-mbl.root_hash.txt.sha256.base64
  mbl-console-image-imx7s-warp-mbl-20181224095924.testdata.json                         mbl-console-image-imx7s-warp-mbl.testdata.json
  mbl-console-image-imx7s-warp-mbl-20181224100038.rootfs.root_hash.txt.sha256.base64    mbl-console-image-imx7s-warp-mbl.verity_header_information.txt
  jonhai01@e113079-lin:/data/work/mbed-linux/master7/build-mbl$
  ```
  Note that the .rootfs.root_hash.txt.sha256.base64 file (created by `do_sign_root_hash`) and the .rootfs.root_hash.txt file (created by `do_generate_verity_metadata`) have different timestamps in their filenames.
